packages/typespec-test/test/batch_modular/generated/typespec-ts/src/api/pagingHelpers.ts

```diff
@@ -0,0 +1,210 @@
+  TElement,
+  TResponse extends PathUncheckedResponse = PathUncheckedResponse
+>(
+  client: Client,
+  getInitialResponse: () => PromiseLike<TResponse>,
+  processResponseBody: (result: TResponse) => Promise<unknown>
+): PagedAsyncIterableIterator<TElement> {
+  let firstRun = true;
+  let itemName: string, nextLinkName: string | undefined;
+  const firstPageLinkPlaceholder = "";
+  const pagedResult: PagedResult<TElement[]> = {
+    firstPageLink: firstPageLinkPlaceholder,
+    getPage: async (pageLink: string) => {
+      const result =
+        firstRun && pageLink === firstPageLinkPlaceholder
+          ? await getInitialResponse()
+          : await client.pathUnchecked(pageLink).get();
+      if (firstRun) {
+        const pageInfo = getPaginationProperties(result);
+        itemName = pageInfo.itemName;
+        nextLinkName = pageInfo.nextLinkName;
+      }
+      firstRun = false;
+      checkPagingRequest(result);
+      const results = await processResponseBody(result as TResponse);
+      const nextLink = getNextLink(results, nextLinkName);
+      const values = getElements<TElement>(results, itemName);
+      return {
+        page: values,
+        nextPageLink: nextLink,
+      };
+    },
+    byPage: (settings?: PageSettings) => {
+      const { continuationToken } = settings ?? {};
+      return getPageAsyncIterator(pagedResult, {
+        pageLink: continuationToken,
+      }) as any;
+    },
+  };
+  const iter: CorePagedAsyncIterableIterator<TElement> =
+    getPagedAsyncIterator(pagedResult);
+
+  return {
+    next() {
+      return iter.next();
+    },
+    [Symbol.asyncIterator]() {
+      return this;
+    },
+    byPage: (settings?: PageSettings) => {
+      return iter.byPage(
+        settings as CorePageSettings
+      ) as unknown as AsyncIterableIterator<
+        TElement[] & { continuationToken?: string }
+      >;
+    },
+  };
+}
+
+async function* getPageAsyncIterator<TPage>(
+  pagedResult: PagedResult<TPage>,
+  options: {
+    pageLink?: string;
+  } = {}
+): AsyncIterableIterator<TPage & { continuationToken?: string }> {
+  const { pageLink } = options;
+  let response = await pagedResult.getPage(
+    pageLink ?? pagedResult.firstPageLink
+  );
+  if (!response) {
+    return;
+  }
+  (response.page as any).continuationToken = response.nextPageLink;
+  yield response.page as any;
+  while (response.nextPageLink) {
+    response = await pagedResult.getPage(response.nextPageLink);
+    if (!response) {
+      return;
+    }
+    (response.page as any).continuationToken = response.nextPageLink;
+    yield response.page as any;
+  }
+}
+
+/**
+ * Gets for the value of nextLink in the body
+ */
+function getNextLink(body: unknown, nextLinkName?: string): string | undefined {
+  if (!nextLinkName) {
+    return undefined;
+  }
+
+  const nextLink = (body as Record<string, unknown>)[nextLinkName];
+
+  if (typeof nextLink !== "string" && typeof nextLink !== "undefined") {
```

Comment:
previously we used the core-paing helper so the firstPage is required with string. So I put a placeholder for that value, but I think this could be changed because Jeff suggested us to build our own helper to avoid cascadings.